core/src/main/java/org/elasticsearch/cluster/routing/allocation/DecisionUtils.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A class that holds utility methods for serializing and working with allocation decisions.
+ */
+public class DecisionUtils {
```

Comment:
I contemplated this.  Given our discussion on the node rebalancing output, I realized that `NodeRebalanceResult` is not needed - all the info is in `NodeAllocationResult`.  Likewise, after iterations on this PR, I realized that an `AbstractAllocationDecision` can encompass a few common elements between all the decisions, so I've done quite a bit of refactoring there.  I also made `NodeAllocationResult` implement `Comparable`.  I implemented all this in 553eb77677fe696fb7d6b314bfed2fe4392a727b.  WDYT?